src/include/common/macros.h

```diff
@@ -14,12 +14,46 @@
 
 #include <cassert>
 #include <exception>
+#include <sstream>
 #include <stdexcept>
 
 namespace bustub {
 
 #define BUSTUB_ASSERT(expr, message) assert((expr) && (message))
 
+namespace internal {
+
+// An internal stream, used to take C++ stream-style parameters for display, and exit the program with `std::abort`.
+class LogFatalStream {
+ public:
+  LogFatalStream(const char *file, int line) : file_(file), line_(line) {}
+
+  ~LogFatalStream() {
+    std::cerr << file_ << ":" << line_ << ": " << log_stream_.str() << std::endl;
+    std::abort();
+  }
+
+  template <typename T>
+  LogFatalStream &operator<<(const T &val) {
+    log_stream_ << val;
+    return *this;
+  }
+
+ private:
+  const char *file_;
+  int line_;
+  std::ostringstream log_stream_;
+};
+
+}  // namespace internal
+
+// A macro which checks `expr` value and performs assert.
+// Different from `BUSTUB_ASSERT`, it takes stream-style parameters.
+#define BUSTUB_ASSERT_AND_LOG(expr)             \
+  if (bool val = (expr); !val) LogFatalStream { \
```

Comment:
If you think necessary, I could add a `Unlikely` function, which improves branch prediction, since I suspect all assertions sits on critical path.